R/utils.R

```diff
@@ -84,3 +55,199 @@ with_unclassed_to <- function(x, .class = 1:length(class(x)), expr,
+    assign("cnds_seek", append(cnds_seek, cnd_seek), envir = envir)
+    assign("cnds", append(cnds, list(cnd)), envir = envir)
+  }
+
+  n_calls <- length(sys.calls())
+  fn_env <- environment()
+  sink(log_file_con, split = split)
+  res <- withVisible(try(withCallingHandlers(
+    if (!quoted) eval(expr_quote, env) else eval(expr, env),
+    condition = function(cnd) {
+      append_cnd(cnd, fn_env)
+      if (inherits(cnd, "message") || inherits(cnd, "warning")) {
+        invokeRestart(computeRestarts()[[1]])
+      } else if (inherits(cnd, "error")) {
+        # trim call stack back to just the scope of the evaluated expression
+        syscalls <- utils::head(utils::tail(sys.calls(), -(9L + n_calls)), -2L)
+        assign("cnds_err_traceback", syscalls, envir = fn_env)
+      }
+    }), silent = TRUE))
+
+  # read as raw so that we can keep carriage return and console-overwrites
+  sink(NULL)
+  close(log_file_con)
+  log_text <- rawToChar(readBin(log_file, "raw", file.size(log_file)))
+  log_text_line_nchars <- nchar(strsplit(gsub("\r", "\n", log_text), "\n")[[1]])
+
+  # NOTE: Windows might use two newline characters "\r\n"?
+  log_newlines <- cumsum(log_text_line_nchars + 1L)
+
+  # rejoin into singular string to split at newlines, as well as any condition
+  # emission points
+  log_cuts <- sort(unique(c(log_newlines, cnds_seek)))
+  log_cuts <- log_cuts[log_cuts < nchar(log_text)]
+  log_text <- substring(log_text, c(1, log_cuts + 1L), c(log_cuts, nchar(log_text)))
+  log_chars <- cumsum(nchar(log_text))
+
+  # find where to insert emitted conditions among output
+  cnd_i <- findInterval(cnds_seek, log_chars)
+  cnds_new_index <- cnd_i + seq_along(cnd_i)
+
+  # inject conditions throughout console output as they were emitted
+  outputs <- rep(list(NULL), length(log_text) + length(cnds_new_index))
+  if (length(cnds_new_index) > 0L) {
+    outputs[cnds_new_index] <- cnds
+    outputs[-cnds_new_index] <- log_text
+  } else {
+    outputs <- log_text
+  }
+
+  any_output_error <- any(vapply(outputs, inherits, logical(1L), "error"))
+
+  structure(
+    if (!quoted) expr_quote else expr,
+    value = res$value,
+    visible = res$visible,
+    traceback = cnds_err_traceback,
+    output = outputs[nzchar(outputs)],
+    class = c("expr_output", "expression"))
+}
+
+
+
+#' Handle pretty printing of expression output
+#'
+#' @param x expr_output to print
+#' @param cr a \code{logical} indicating whether carriage returns should be
+#'   printed, possibly overwriting characters in the output.
+#' @param ... additional arguments unused
+#' @param sleep an \code{numeric} indicating a time to sleep between printing
+#'   each line to console. This can be helpful if the original output overwrites
+#'   valuable information in the log that is eventually overwritten and you
+#'   would like to watch it play out as it was formatted.
+#'
+#' @export
+#'
+#'
+print.expr_output <- function(x, cr = TRUE, ..., sleep = 0) {
```

Comment:
How will we need this function? I did not see it is used in other place. 

Shall the name be `print_expr_output`? Otherwise, would you plan to create a class called "expr_output" to use `print` method?

Reply:
Right now it's unused. In the future, if we want to capture output during metric evaluation and print it out to an audit log, we can use this to reformat the output as it would have appeared if the code was run in an interactive session. 
I'll add in an example of what this might look like, but if you want to try it out you can take a look at

```r
# with the ./riskmetric project directory as your working directory

console_output <- riskmetric:::capture_expr_output({
  # output i's, looping through output type
  for (i in 0:10) {
    f <- switch(i %% 3 + 1L, print, message, warning)
    f(i)
  }
})

print(console_output)
# > for (i in 0:10) {
#   f <- switch(i%%3 + 1L, print, message, warning)
#   f(i)
# }
# [1] 0
# 1
# Warning in eval(expr_quote, env): 2
# [1] 3
# 4
# Warning in eval(expr_quote, env): 5
# [1] 6
# 7
# Warning in eval(expr_quote, env): 8
# [1] 9
# 10
```

You can also check out some of the attributes to explore the information we capture during evaluation. 

It also needs to be called `print.<object class>` so that the print function dispatches to the appropriate function for formatting.